tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -226,6 +226,296 @@ static bool HasCanonicalDimensionNumbers(
+                                                          out_ein};
+
+    if (!CheckBatchHasEqualRank(lhs_ein.size(), lhs_loop, rhs_ein.size(),
+                                rhs_loop, out_ein.size(), out_loop)) {
+      return rewriter.notifyMatchFailure(
+          op, "Invalid elipsis('...') within einsum config!");
+    }
+
+    // Find all unique indices in the input and output.
+    llvm::SmallSetVector<StringRef, 4> input_ind;
+    llvm::SmallSetVector<StringRef, 4> output_ind;
+    for (const auto& loops_it : llvm::enumerate(all_loop_dim)) {
+      for (const StringRef& dim : loops_it.value()) {
+        if (loops_it.index() < args.size()) {
+          input_ind.insert(dim);
+        } else {
+          output_ind.insert(dim);
+        }
+      }
+    }
+
+    llvm::SmallSetVector<StringRef, 4> reduction_axe =
+        FindSummationAxes(input_ind, output_ind);
+
+    // Find input/output values and types.
+    auto loc = op.getLoc();
+
+    // Prepare init tensor for linalg.generic op.
+    auto dyn_sizes = ExtractDynamicEinsumSizes(
+        rewriter, loc, adaptor.lhs(), adaptor.rhs(), lhs_ein, rhs_ein, out_ein);
+    Value output = GetInitTensor(rewriter, loc, result_ty, dyn_sizes);
+    if (reduction_axe.size() > 0) {
+      auto zero_attr = rewriter.getZeroAttr(result_ty.getElementType());
+      Value zero = rewriter.create<ConstantOp>(loc, zero_attr);
+      output = rewriter.create<linalg::FillOp>(loc, zero, output).getResult(0);
+    }
+
+    // Create indexing maps.
+    // Create a 1:1 map from f:strDimension -> affineDimension.
+    int64_t nloops = input_ind.size();
+    DenseMap<StringRef, AffineExpr> str_affine_dim_umap;
+    for (auto it : llvm::enumerate(input_ind)) {
+      str_affine_dim_umap[it.value()] = rewriter.getAffineDimExpr(it.index());
+    }
+
+    // From einsum_config of each operand in vector<string>, generate
+    // the equivalent vector<AffineExpr>.
+    SmallVector<AffineMap, 4> maps;
+    for (const auto& loop_operand : all_loop_dim) {
+      auto exprs = GetExprFromConfig(loop_operand, str_affine_dim_umap);
+      maps.push_back(AffineMap::get(nloops, 0, exprs, rewriter.getContext()));
+    }
+
+    auto linalg_op = rewriter.create<linalg::GenericOp>(
+        loc, result_ty ? result_ty : TypeRange{}, args, output, maps,
+        GetEinsumLoopsAttrs(input_ind, reduction_axe),
+        [&](OpBuilder& b, Location nested_loc, ValueRange args) {
+          Value result_val =
+              b.create<mlir::MulFOp>(nested_loc, args[0], args[1]);
+          ;
+          if (reduction_axe.size() > 0) {
+            result_val =
+                b.create<mlir::AddFOp>(nested_loc, args[2], result_val);
+          }
+          b.create<linalg::YieldOp>(nested_loc, result_val);
+        });
+    rewriter.replaceOp(op, linalg_op.getResults());
+    return success();
+  }
+
+ private:
+  static const std::string kArrow_;
+  static const std::string kComma_;
+  static const std::string kEllipsis_;
```

Comment:
Sorry, it looks like the style guide like using lower case letters separated by '_' characters with a '_' character at the end, e.g. a_class_data_member_.

also, static and global variables are forbidden unless they are trivially destructible, so maybe just mark them static?

Reply:
done. Though I have some questions on it, is it considered "global" even though it's a private member in a class? If that's the case why do we need to drop the "const"(since we are only marking it as just static) to fix the "globalness" of it?

Reply:
Please use `static constexpr StringRef` here. 

These are global entities of the program, and there will be a constructor executed on program startup (and a destructor on exit).

Reply:
thanks for the explanation @joker-eph , done :)